upgrade.go

```diff
@@ -0,0 +1,604 @@
+// with -test.v to see the data logged.
+func (u *UpgradeCoverage) Report(t *testing.T) {
+	t.Run("coverage", func(t *testing.T) {
+		covered := u.resources
+		t.Logf("Resources covered: %d", len(covered))
+
+		sorted := []string{}
+		for k := range covered {
+			sorted = append(sorted, k)
+		}
+		sort.Strings(sorted)
+		for _, s := range sorted {
+			t.Logf("- %s", s)
+		}
+	})
+}
+
+// Enumerates various available modes to test provider upgrades. The modes differ in speed vs
+// precision tradeoffs.
+type UpgradeTestMode int
+
+const (
+	// The least precise but fastest mode. Tests are performed in-memory using pre-recorded
+	// snapshots of baseline provider behavior. No cloud credentals are required, no
+	// subprocesses are launched, fully debuggable.
+	UpgradeTestMode_Quick UpgradeTestMode = iota
+
+	// The medium precision/speed mode. Imports Pulumi statefile recorded on a baseline version,
+	// and performs pulumi preview, assrting that the preview results in an empty diff. Cloud
+	// credentials are required, but no infra is actually provisioned, making it quicker to
+	// verify slow-to-provision resources such as databases.
+	UpgradeTestMode_PreviewOnly
+
+	// Full fidelity, slow mode. No pre-recorded snapshots are required. Do a complete pulumi up
+	// on the baseline version, followed by a complete pulumi up on the version under test, and
+	// assert that there are no observable updates or replacements.
+	UpgradeTestMode_Full
+)
+
+func (m UpgradeTestMode) String() string {
+	switch m {
+	case UpgradeTestMode_PreviewOnly:
+		return "PreviewOnly"
+	case UpgradeTestMode_Quick:
+		return "Quick"
+	case UpgradeTestMode_Full:
+		return "Full"
+	}
+	return "<unknown>"
+}
+
+func WithSkippedUpgradeTestMode(m UpgradeTestMode, reason string) Option {
+	contract.Assertf(reason != "", "reason cannot be empty")
+	contract.Assertf(m.String() != "<unknown>", "unknown UpgradeTestMode")
+	return func(b *ProviderTest) {
+		if b.upgradeOpts.modes == nil {
+			b.upgradeOpts.modes = map[UpgradeTestMode]string{}
+		}
+		b.upgradeOpts.modes[m] = reason
+	}
+}
+
+func WithBaselineVersion(v string) Option {
+	contract.Assertf(v != "", "BaselineVersion cannot be empty")
+	return func(b *ProviderTest) { b.upgradeOpts.baselineVersion = v }
+}
+
+func WithProviderName(name string) Option {
+	contract.Assertf(name != "", "ProviderName cannot be empty, "+
+		"expecting a provider name like `gcp` or `aws`")
+	return func(b *ProviderTest) { b.upgradeOpts.providerName = name }
+}
+
+func WithResourceProviderServer(s pulumirpc.ResourceProviderServer) Option {
+	contract.Assertf(s != nil, "ResourceProviderServer cannot be nil")
+	return func(b *ProviderTest) { b.upgradeOpts.resourceProviderServer = s }
+}
+
+func WithUpgradeCoverage(uc *UpgradeCoverage) Option {
+	contract.Assertf(uc != nil, "UpgradeCoverage cannot be nil")
+	return func(b *ProviderTest) { b.upgradeOpts.upgradeCoverage = uc }
+}
```

Comment:
I guess this method is why UpgradeCoverage is exported, but this method could just new up the instance right? I don't follow why the user of this library would pass it in as an option.

Reply:
Killed it.